### PR TITLE
mgr/dashboard: add scss linting

### PR DESCRIPTION
**Pull request by ishanrai05**
Fixes: https://tracker.ceph.com/issues/46108

## Checklist
- [x] References tracker ticket
- [x] Updates documentation if necessary
- [x] Includes tests for new functionality or reproducer for bug

---

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard backend`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`

</details>

### CODE REVIEW

**Comment by Devp00l** [comment]
To which tracker issue are you referring? Could you please add it to your PR description and to your commit message?

**Comment by ishanrai05** [comment]
@Devp00l Hi. I don't think there is a tracker issue directly associated with this. This was suggested while working on https://tracker.ceph.com/issues/38891. Should I add it?

**Comment by ishanrai05** [comment]
jenkins test dashboard

**Comment by ishanrai05** [comment]
jenkins test dashboard backend

**Comment by LenzGr** [comment]
> This was suggested while working on [tracker.ceph.com/issues/38891](https://tracker.ceph.com/issues/38891). Should I add it?

Would you mind creating a new feature tracker issue for this one, @ishanrai05 ?
Then it should be marked as related to issue 38891.

**Comment by LenzGr** [comment]
Is there anything that should be added to `HACKING.rst` about these new checks?

**Comment by ishanrai05** [comment]
@LenzGr yes, I think we should add `npm run scss:lint` to Hacking.rst as a part of linting

**Comment by ishanrai05** [comment]
@tspmelo I can update the codebase according to `stylelint-config-sass-guideline` one guideline at a time. I avoided it in this PR because there would be lots of changes required. Should I do it in this PR itself?

**Comment by tspmelo** [comment]
We can discuss this in tomorrow's standup .

**Comment by tspmelo** [comment]
As discussed, you should add a new commit to this PR with the usage of `stylelint-config-sass-guideline`.

File: src/pybind/mgr/dashboard/frontend/.stylelintrc [hunk 1]
```diff
@@ -0,0 +1,36 @@
+{
+  "extends": "stylelint-config-sass-guidelines",
+  "plugins": [
+    "stylelint-declaration-use-variable"
+  ],  
+  "rules": {
+    "declaration-property-value-blacklist":null,
+    "function-parentheses-space-inside":null,
+    "indentation":null,
```
Comment (ishanrai05) [inline src/pybind/mgr/dashboard/frontend/.stylelintrc:9]: `function-parentheses-space-inside` and `indentation` cause conflict with `npm run fix:prettier` and hence disabled,

**Comment by ishanrai05** [comment]
```
src/styles/defaults.scss
186:5  ✖  Unnecessary nesting selector (&)   scss/selector-no-redundant-nesting-selector
```

I am not sure how to correct this. @tspmelo

**Comment by tspmelo** [comment]
@ishanrai05  I think you can remove  the `& {` and the corresponding `}`

**Comment by ishanrai05** [comment]
jenkins test dashboard backend

**Comment by tspmelo** [comment]
jenkins test docs

File: src/pybind/mgr/dashboard/frontend/package.json [hunk 2]
```diff
@@ -26,11 +26,13 @@
     "lint:prettier": "prettier --list-different \"{src,cypress}/**/*.{ts,scss}\"",
     "lint:html": "htmllint src/app/**/*.html && html-linter --config html-linter.config.json",
     "lint:tsc": "tsc -p src/tsconfig.app.json --noEmit && tsc -p tsconfig.spec.json --noEmit && tsc -p cypress/tsconfig.json --noEmit",
-    "lint": "npm run lint:tsc && npm run lint:tslint && npm run lint:prettier && npm run lint:html",
+    "lint:scss": "stylelint '**/*.scss'",
+    "lint": "npm run lint:tsc && npm run lint:tslint && npm run lint:prettier && npm run lint:html && npm run lint:scss",
```
Comment (epuertat) [inline src/pybind/mgr/dashboard/frontend/package.json:30]: I reiterate this suggestion: https://github.com/ceph/ceph/pull/35225#discussion_r433164558

File: src/pybind/mgr/dashboard/frontend/package.json [hunk 3]
```diff
@@ -26,11 +26,13 @@
...
     "fix:prettier": "prettier --write \"{src,cypress}/**/*.{ts,scss}\"",
     "fix:tslint": "npm run lint:tslint -- --fix",
+    "fix:scss": "stylelint '**/*.scss' --fix",
     "fixmod": "pretty-quick --pattern \"{src,cypress}/**/*.{ts,scss}\" --branch HEAD",
-    "fix": "npm run fix:tslint; npm run fix:prettier",
+    "fix": "npm run fix:tslint; npm run fix:prettier; npm run fix:scss",
```
Comment (epuertat) [inline src/pybind/mgr/dashboard/frontend/package.json:35]: And same for here: https://github.com/ceph/ceph/pull/35225#discussion_r433164558

File: src/pybind/mgr/dashboard/frontend/package.json (hunk 2 above)
Comment (ishanrai05) [reply src/pybind/mgr/dashboard/frontend/package.json]: done.

File: src/pybind/mgr/dashboard/frontend/package.json (hunk 3 above)
Comment (ishanrai05) [reply src/pybind/mgr/dashboard/frontend/package.json]: done. thank you

File: src/pybind/mgr/dashboard/frontend/package.json [hunk 4]
```diff
@@ -121,9 +123,13 @@
     "ng-bullet": "1.0.3",
     "ng-mocks": "9.5.0",
     "npm-force-resolutions": "0.0.3",
+    "npm-run-all": "^4.1.5",
```
Comment (tspmelo) [inline src/pybind/mgr/dashboard/frontend/package.json:126]:
```suggestion
    "npm-run-all": "4.1.5",
```

Comment (ishanrai05) [reply src/pybind/mgr/dashboard/frontend/package.json]: done

**Comment by ishanrai05** [comment]
jenkins test docs

**Comment by tspmelo** [comment]
jenkins test docs

**Comment by ishanrai05** [comment]
jenkins test dashboard backend

**Comment by ishanrai05** [comment]
jenkins test dashboard

**Comment by ishanrai05** [comment]
jenkins test dashboard

**Comment by ishanrai05** [comment]
jenkins test dashboard backend

**Comment by ishanrai05** [comment]
jenkins test dashboard

**Pull request merged**